File: December2021/StreamOfCharacters.java
class StreamChecker {
	private class TrieNode{
		private TrieNode[] children =null;
		private boolean isLeaf;
		public TrieNode(){
			children = new TrieNode[26];
		}
	}

	private TrieNode root = null;

	private StringBuilder queryStr = null;

	public StreamChecker(String[] words) {
		root = new TrieNode();
		queryStr = new StringBuilder();
		for(String word : words) {
			addWord(word);
		}
	}


	private void addWord(String word){
		TrieNode it = root;
		for(int i=word.length()-1;i>=0;i--){ // Iterating in revese order
			char c = word.charAt(i);
			int index = c - 'a';
			if(it!=null && it.children[index]==null){
				TrieNode newNode = new TrieNode();
				it.children[index] = newNode;
			}
			it = it.children[index];
		}
		it.isLeaf = true;
	}

	public boolean query(char letter) {
		queryStr.append(letter);
		return search(queryStr.toString());
	}

	private boolean search(String queryStr){
		TrieNode it = root;
		for(int i=queryStr.length()-1;i>=0;i--){
			char c = queryStr.charAt(i);
			int index = c - 'a';

			if(it!=null && it.children[index]!=null){
				it = it.children[index];
				if(it.isLeaf){
					return true;
				}
			} else {
				return false;
			}
		}
		return false;
	}

}

File: December2021/C++/Stream_of_Characters.cpp
class Trie{
   
    public :
        Trie()
        {
            for(auto& i:children)
                i=NULL;
            end=false;
        }
        Trie* children[26];
        bool end;
    
};
class StreamChecker {
    Trie* trie;
    deque<char> q;
    
public:
    StreamChecker(vector<string>& words) {
        trie=new Trie();
        for(auto& i:words)
        {
            reverse(begin(i),end(i));
            insert(i,trie);
        }
        
    }
    

    
    bool query(char letter) {
        q.push_front(letter);
    
        Trie* curr=trie;
        for(int i=0;i<q.size();i++)
        {
            int idx=q[i]-'a';
            curr=curr->children[idx];
            if(!curr) return false;
            if(curr->end)
                return true;
            
        }
        return curr->end;
    }
    
    void insert(string word,Trie* root)
    {
        Trie* node=root;
        for(int i=0;i<word.size();i++)
        {
            int idx=word[i]-'a';
            if(node->children[idx]==NULL)
                node->children[idx]=new Trie();
            node=node->children[idx];
        }
        
        node->end=true;
    }
};

/**
 * Your StreamChecker object will be instantiated and called as such:
 * StreamChecker* obj = new StreamChecker(words);
 * bool param_1 = obj->query(letter);
 */